### ISSUE TITLE
Validation of Models on Loss Function

### ISSUE THREAD

**Issue by tsafs** [hfawaz/dl-4-tsc#3]
Hello

Thanks again for sharing your code.

I successfully applied your code on my datasets and found, when inspecting the results in e.g. `df_best_model.csv` and `df_metrics`, that you deem the model as the best model, by monitoring the loss function of the training set. For example, in the FCN model you are setting a ModelCheckpoint by monitoring the training loss. But in the mentioned metrics and your paper you rate the model by its validation accuracy, is that correct?  

```python
file_path = self.output_directory+'best_model.hdf5'

model_checkpoint = keras.callbacks.ModelCheckpoint(filepath=file_path, 
    monitor='loss', save_best_only=True)

self.callbacks = [reduce_lr,model_checkpoint]

hist = self.model.fit(x_train, y_train, batch_size=mini_batch_size, 
    epochs=nb_epochs, verbose=self.verbose, validation_data=(x_val,y_val), 
    callbacks=self.callbacks)
```

I have run through multiple model / dataset variations and found that sometimes Model A with parameters A is better, measured in training loss, than Model A with parameters B. However, in these cases Model A with parameters A overfitted and its validation loss is worse than that of Model A with parameters B. When measuring the performance of a model by its validation accuracy, wouldn't it be more accurate to set the ModelCheckpoint to monitor `val_loss` or `val_acc` instead of `train_loss`?

Thanks

**Comment by patrickzib** [comment]
Generally speaking: this is correct. If you have access to a train, validation and test split, you can use the validation split to measure the performance and to set ModelCheckpoint to `val_los` or` val_acc`. In the end you evaluation your model using the held-back test data

However, you might have notice that the UCR datasets consist of only train and test datasets. Furthermore, the train data is often too small (with some dozens of samples) to split it into train and validation sets. Thus, this code skips the validation set and passes the `test data` as validation data to self.model.fit(). Thus, to avoid any bias in model selection, you have to use `train_loss` - otherwise you would select your model using the accuracy on the test set.

**Comment by hfawaz** [comment]
Hello,

Thank you @sebastianfast for your interest and feedback on our work. 

We are glad that the community is founding it useful for their research!

Indeed, choosing a model based on its training loss accuracy may not be the best solution.
However this was not our choice, in fact since we are re-implementing other approaches we followed the same methodology in their corresponding original paper.

For example for FCN, in their original paper the authors did choose the best model based on the training loss which is why we did the same thing when re-implementing FCN in our framework.
Furthermore in the paper, we specified for which models we used the training loss and for which ones we used a validation loss (by taking a non-seen split from the train set).
Column "valid" in Table 2 from our paper lists two values: either "train" which means the model was selected on the train loss or "valid-x%" which means the that the model was selected on a validation set split (x%) from the train set.

I hope the issue is clear now.

Let me know if you have any other questions.

@patrickzib Thank you for responding to the question ! 

Best regards,

Hassan

**Comment by tsafs** [comment]
Hallo @patrickzib and  @hfawaz,

Thanks for responding in such a short time to both of you. It definitely made things clearer and I must admit that I overread the mentions in the paper of you validating the networks based on the training loss.
